### PR TITLE
feat: enable balancer phase for plugins

### PR DESCRIPTION
**Pull request by spacewander**
Signed-off-by: spacewander <spacewanderlzx@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Review by tokers — APPROVED** [review]

**Comment by Yiyiyimu** [comment]
cc @dmsolr

**Review by tzssangglass — APPROVED** [review]

**Review by gxthrj — APPROVED** [review]
LGTM

**Review by dmsolr — APPROVED** [review]
LGTM

File: apisix/balancer.lua [hunk 1]
```diff
@@ -308,15 +309,24 @@ function _M.run(route, ctx)
...
                 -- retried node has a different host
                 ctx.var.upstream_host = host
-                balancer.recreate_request()
+                header_changed = true
             end
         end
+
+    end
+
+    local _, run = plugin_funcs("balancer")
```
Comment (starsz) [inline apisix/balancer.lua:326]: We will return balancer twice?
Line 312 and line 326 ?

**Review by tao12345666333 — APPROVED** [review]

File: apisix/balancer.lua (hunk 1 above)
Comment (spacewander) [reply apisix/balancer.lua:324]: Fixed in the new commit

**Review by starsz — APPROVED** [review]

File: t/plugin/serverless.t [hunk 2]
```diff
@@ -679,3 +679,159 @@ GET /hello
...
+
+
+
+=== TEST 28: check plugin
+--- request
+GET /log_request
+--- grep_error_log eval
+qr/(run balancer phase with [\d.]+)/
+--- grep_error_log_out
+run balancer phase with 127.0.0.2
```
Comment (tzssangglass) [inline t/plugin/serverless.t:836]: I don't understand why there are two `run balancer phases with ……`

Comment (spacewander) [reply t/plugin/serverless.t:836]: The last one is for retry

**Pull request merged**